### PR TITLE
Better no same operation

### PR DESCRIPTION
**Pull request by joonazan**

### CODE REVIEW

File: server/cache/operator.go [hunk 1]
```diff
@@ -33,14 +40,39 @@ func (o Operator) GetBlob(operations ...ops.Operation) (blob images.ImageBlob, e
 	if start == len(operations) {
 		return startimage, nil
 	} else {
-		t := <-o.tokens
-		defer func() { o.tokens <- t }()
+		o.Lock()
+		cond, inProgress := o.inProgress[key]
+		if !inProgress {
+			// image may have entered cache while this goroutine moved to this place in code
```
Comment (piiroim1) [inline server/cache/operator.go]: Exactly in what situation?

**Comment by joonazan** [comment]
It is theoretically possible that thread A does not find an image, thread B then caches it and removes in from inProgress which causes A to generate the image once more.

This could probably be avoided by doing synchronization in a more sensible way, but I can't figure out how. I think even the rarest cases should be accounted for. The problem with this one is that it is almost untestable.

The other rare special case can easily be reproduced. I should actually write a test for it unless someone figures out how to beautify this code.

File: server/cache/operator.go [hunk 2]
```diff
@@ -33,14 +40,39 @@ func (o Operator) GetBlob(operations ...ops.Operation) (blob images.ImageBlob, e
...
+			o.Lock()
+			o.addInProgress(key)
+			o.Unlock()
 		}
 
+		o.tokens.Borrow()
+		defer o.tokens.Return()
+
 		img := images.NewImage()
 		defer img.Destroy()
```
Comment (piiroim1) [inline server/cache/operator.go]: IMO start and startimage should be fetched from cache here

**Comment by piiroim1** [comment]
I added some more comments. I also think that splitting GetBlob() to multiple sub-functions would make the code more readable/undersandable - if the functions can be named descriptively.

**Comment by joonazan** [comment]
Actually inProgress should be checked for each partial solution after every cache check. I guess I have to upgrade the lock to an RWMutex.

Oh, and inProgress should be a slice, but that is a minimal performance loss because for checking cache expensive operations are needed.

Do you agree that the Cacher should actually just take keys, not []ops?

**Comment by piiroim1** [comment]
Actually if we are going to do sub-resizes, it is actually possible that some suboperation is inProgress. So yes, inProgress check should be done for suboperations as well.

Cache should take just keys so there is no need to generate them separately for operator and cache.

I'll merge this and these improvements can be worked on later.

**Pull request closed without merging**